mirgecom/mpi.py

```diff
@@ -117,6 +117,8 @@ def _check_gpu_oversubscription() -> None:
     if size <= 1:
         return
 
+    # This may unnecessarily require pyopencl in case we run with a
+    # NumpyArrayContext or CupyArrayContext
     cl_ctx = cl.create_some_context()
     dev = cl_ctx.devices
```

Comment:
Creating an OpenCL context is tremendously expensive, and doing so just to find the device to be used is not a good idea. Instead of this, there should be a central place in mirgecom that decides on the target device, and that should be used instead.

To support this, [`create_some_context` in pyopencl](https://github.com/inducer/pyopencl/blob/4e067bdd3ee06c164572c4e9ae398beee6b58663/pyopencl/__init__.py#L1555) should be refactored to separate device choice from context creation, if that device choice logic is intended to be retained.

Reply:
Is https://github.com/inducer/pyopencl/pull/735 what you had in mind?

Reply:
Note that #1030 goes a bit in a different direction - it centralizes actx creation in mpi_entry_point and removes all other calls to create_some_context.